JoodHajeer/assignment4/Source.cpp

```diff
@@ -0,0 +1,55 @@
+#include <iostream>
+#include <stdio.h>  
+#include <string>
+#include <queue>
+#include <set>
+#include <stdlib.h>
+#include"codeu_test_lib.h"
+using namespace std;
+
+void DFS(int curRow, int curCol, int numberOfRows, int numberOfColumns, vector<vector<bool> > & tilesMap) {
+	tilesMap[curRow][curCol] = false; //set this tile visited
+	int dr[] = {0,0,1,-1};
+	int dc[] = {1,-1,0,0};
+	for (int i=0;i<4;i++){
+		int newRow = curRow + dr[i];
+		int newCol = curCol + dc[i];
+		if (newRow < 0 || newRow >= numberOfRows || newCol < 0 || newCol >= numberOfColumns || !tilesMap[newRow][newCol]) // skip tiles that are out on bound or water or already visited 
+			continue;
+		DFS(newRow, newCol, numberOfRows, numberOfColumns, tilesMap);
+	}
+}
+int countIslands(int numberOfRows, int numberOfColumns, vector<vector<bool> > tilesMap) {
```

Comment:
The copy of tilesMap is intentional, right? Nice :)
Note: numberOfColumns & numberOfRows are redundant, they can be extracted from tilesMap. Also, minor thing, but the rows of tilesMap may have different sizes.